simSPI/linear_simulator/params_utils.py

```diff
@@ -0,0 +1,361 @@
+        angle_tilt = np.array(self.particle["rlnAngleTilt"], ndmin=2).squeeze()
+        angle_rot = np.array(self.particle["rlnAngleRot"], ndmin=2).squeeze()
+        relion_euler_np = np.radians(
+            np.stack(
+                [
+                    -angle_psi,  # convert Relion to our convention
+                    angle_tilt * (-1 if self.config.relion_invert_hand else 1),
+                    # convert Relion to our convention + invert hand
+                    angle_rot,
+                ]
+            )
+        )  # convert Relion to our convention
+        relion_euler = torch.from_numpy(relion_euler_np).t()  # Bx3
+        rotmat = euler_angles_to_matrix(relion_euler, convention="ZYZ")
+
+        return {
+            "rotmat": rotmat,
+            "relion_angle_psi": angle_psi,
+            "relion_angle_tilt": angle_tilt,
+            "relion_angle_rot": angle_rot,
+        }
+
+    def get_ctf_params(self):
+        """Get the parameters for the CTF of the particle from the starfile.
+
+        If config.ctf is True else returns {}
+
+        Returns
+        -------
+        ctf_params: dict of type str to {tensor}
+            dictionary containing:
+            "defocus_u": torch.Tensor
+                Tensor of size (batch_size,1,1,1) that contains the major
+                defocus value in microns
+            "defocus_v": torch.Tensor
+                Tensor of size (batch_size,1,1,1) that contains the minor
+                defocus value in microns
+            "defocus_angle": torch.Tensor
+                Tensor of size (batch_size,1,1,1) that contains the major
+                defocus value in microns
+        """
+        params = None
+        if self.config.ctf:
+            defocus_u = (
+                torch.from_numpy(np.array(self.particle["rlnDefocusU"] / 1e4))
+                .float()
+                .squeeze()[:, None, None, None]
+            )
+            defocus_v = (
+                torch.from_numpy(np.array(self.particle["rlnDefocusV"] / 1e4))
+                .float()
+                .squeeze()[:, None, None, None]
+            )
+            defocus_angle = (
+                torch.from_numpy(np.radians(np.array(self.particle["rlnDefocusAngle"])))
+                .float()
+                .squeeze()[:, None, None, None]
+            )
+            params = {
+                "defocus_u": defocus_u,
+                "defocus_v": defocus_v,
+                "defocus_angle": defocus_angle,
+            }
+        return params
+
+    def get_shift_params(self):
+        """Get the parameters for the shift of the particle from the starfile.
+
+        If config.shift is True else returns {}
+
+        Returns
+        -------
+            shift_params: dict of type str to {torch.Tensor}
+            dictionary containing
+            'shift_x': torch.Tensor (batch_size,)
+                batch of shifts along horizontal axis
```

Comment:
put in the units. pixels, angstroms, etc?

Reply:
Added now. Thanks